dialogue-core/src/main/java/com/palantir/dialogue/core/StickyEndpointChannels2.java

```diff
@@ -0,0 +1,212 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Value.Enclosing
+public final class StickyEndpointChannels2 implements Supplier<Supplier<Channel>> {
+
+    private static final Logger log = LoggerFactory.getLogger(StickyEndpointChannels2.class);
+
+    private final EndpointChannelFactory delegate;
+
+    private StickyEndpointChannels2(EndpointChannelFactory endpointChannelFactory) {
+        this.delegate = Preconditions.checkNotNull(endpointChannelFactory, "endpointChannelFactory");
+    }
+
+    @Override
+    public Supplier<Channel> get() {
+        return new StickySessionSupplier(delegate);
+    }
+
+    @Override
+    public String toString() {
+        return "StickyEndpointChannels2{}";
+    }
+
+    private static final class StickySessionSupplier implements Supplier<Channel> {
+
+        private final EndpointChannelFactory channelFactory;
+        private final StickyRouter router = new DefaultStickyRouter();
+
+        private StickySessionSupplier(EndpointChannelFactory channelFactory) {
+            this.channelFactory = channelFactory;
+        }
+
+        @Override
+        public Channel get() {
+            return new Sticky(channelFactory, router);
+        }
+    }
+
+    private static final class Sticky implements EndpointChannelFactory, Channel {
+
+        private final EndpointChannelFactory channelFactory;
+        private final StickyRouter router;
+        private final RoutingKey routingKey;
+
+        private Sticky(EndpointChannelFactory channelFactory, StickyRouter router) {
+            this.router = router;
+            this.channelFactory = channelFactory;
+            this.routingKey = RoutingKey.create();
+        }
+
+        @Override
+        public EndpointChannel endpoint(Endpoint endpoint) {
+            return new StickyEndpointChannel(router, channelFactory.endpoint(endpoint), routingKey);
+        }
+
+        /**
+         * .
+         * @deprecated prefer {@link #endpoint}, as it allows binding work upfront
+         */
+        @Deprecated
+        @Override
+        public ListenableFuture<Response> execute(Endpoint endpoint, Request request) {
+            return endpoint(endpoint).execute(request);
+        }
+
+        @Override
+        public String toString() {
+            return "Sticky{" + channelFactory + '}';
+        }
+    }
+
+    public static StickyEndpointChannels2 create(EndpointChannelFactory endpointChannelFactory) {
+        return new StickyEndpointChannels2(endpointChannelFactory);
+    }
+
+    interface StickyRouter {
+        ListenableFuture<Response> execute(Request request, EndpointChannel endpointChannel);
+    }
+
+    private static final class DefaultStickyRouter implements StickyRouter {
+
+        private final InFlightCallCallback callback = new InFlightCallCallback();
+
+        @Nullable
+        private volatile HostId hostId;
+
+        @Nullable
+        private volatile ListenableFuture<Response> callInFlight;
```

Comment:
Let's help error-prone help us:
```suggestion
        @Nullable
        @GuardedBy("this")
        private volatile ListenableFuture<Response> callInFlight;
```